src/software/jetson_nano/rotary_encoder/RotaryEncoder.py

```diff
@@ -0,0 +1,123 @@
+import Jetson.GPIO as GPIO
+
+# Channel B and Channel A states
+STATE_0 = [0, 0]
+STATE_1 = [0, 1]
+STATE_2 = [1, 1]
+STATE_3 = [1, 0]
+
+CLOCKWISE = 1
+COUNTERCLOCKWISE = -1
+IDLE = 0
+
+
+class RotaryEncoder:
+    def __init__(
+        self,
+        A_PIN,
+        B_PIN,
+        BUTTON_PIN,
+        on_clockwise_rotate,
+        on_counterclockwise_rotate,
+        on_click,
+    ):
+        self.A_PIN = A_PIN
+        self.B_PIN = B_PIN
+        self.BUTTON_PIN = BUTTON_PIN
+
+        self.on_clockwise_rotate = on_clockwise_rotate
+        self.on_counterclockwise_rotate = on_counterclockwise_rotate
+        self.on_click = on_click
+
+        self.counts_per_rotation = 4
+
+    def init(self):
+        """ Initialize GPIO pins and rotary encoder state """
+        GPIO.setmode(GPIO.BOARD)
+
+        # These pins will need external pullups set up
+        GPIO.setup(self.A_PIN, GPIO.IN)
+        GPIO.setup(self.B_PIN, GPIO.IN)
+        GPIO.setup(self.BUTTON_PIN, GPIO.IN)
+
+        a_state = GPIO.input(self.A_PIN)
+        b_state = GPIO.input(self.B_PIN)
+
+        self.curr_state = [b_state, a_state]
+        self.dir = IDLE
+
+        self.count = 0
+
+    # Channel B and A go through 4 states for a single rotation.
+    # The following are values for clockwise rotation:
+    #  Channel B Channel A
+    #     0          0
+    #     0          1
+    #     1          1
+    #     1          0
+    def rot_state(self):
+        """ FSM to determine rotation direction and count state transitions """
+        a_state = GPIO.input(self.A_PIN)
+        b_state = GPIO.input(self.B_PIN)
+        curr_state = [b_state, a_state]
+
+        # State transitions
+        if self.curr_state == STATE_0:
+            if curr_state == STATE_1:
+                self.dir = CLOCKWISE
+            elif curr_state == STATE_3:
+                self.dir = COUNTERCLOCKWISE
+        elif self.curr_state == STATE_1:
+            if curr_state == STATE_2:
+                self.dir = CLOCKWISE
+            elif curr_state == STATE_0:
+                self.dir = COUNTERCLOCKWISE
+        elif self.curr_state == STATE_2:
+            if curr_state == STATE_3:
+                self.dir = CLOCKWISE
+            elif curr_state == STATE_1:
+                self.dir = COUNTERCLOCKWISE
+        else:
+            if curr_state == STATE_0:
+                self.dir = CLOCKWISE
+            elif curr_state == STATE_2:
+                self.dir = COUNTERCLOCKWISE
```

Comment:
Suggestion for how you could make this is a bit more concise
~~~

# curr_state : next_state  indicates clockwise rotation
clockwise_state_transitions =\
{
     STATE_0: STATE_1
     STATE_1: STATE_2
     STATE_2: STATE_3
     STATE_3: STATE_0
}

# curr_state : next_state  indicates counter clockwise rotation

counter_clockwise_state_transitions =\
{
     STATE_0: STATE_3
     STATE_1: STATE_0
     STATE_2: STATE_1
     STATE_3: STATE_2
}

if clockwise_state_transitions[self.curr_state] == next_state:
     self.on_clockwise_rotate()
elif counterclockwise_state_transitions[self.curr_state] == next_state:
     self.on_counterclockwise_rotate() 
~~~

and then I think this code can just live in `update_state`. If it doesn't let you hash a list, you  can switch to a tuple

Reply:
I added the dictionaries as global variables and added the if statement you suggested to the function that calls the callback functions to more concisely determine rotation direction. I have kept the `rot_state` function to increment the counter to know when a full transition has completed.